sprint-2/step-1 initial commit

piladik opened the pull request.

piladik committed:
sprint-2/step-1 initial commit

gennady-bars commented on src/utils/burger-api.js:
```diff
@@ -1,4 +1,5 @@
 const BURGER_API_URL = "https://norma.nomoreparties.space/api";
+const ORDER_API_URL = "https://norma.nomoreparties.space/api/orders";
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.


на 1й строчке есть уже базовый урл

gennady-bars commented on src/utils/burger-api.js:
```diff
@@ -9,7 +10,7 @@ const checkResponse = (res) => {
+export const getIngredients = async () => {
   const res = await fetch(`${BURGER_API_URL}/ingredients`);
 
   checkResponse(res);
```
**Можно лучше**

Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -58,10 +59,12 @@ function App() {
+        <IngredientsContext.Provider value={{ ingredients, setIngredients }}>
+          <main className="main">
+            <BurgerIngredients />
+            <BurgerConstructor ingredients={data} />
```
В ` BurgerIngredients` и `BurgerConstructor` больше не нужно передавать в `props` массив ингредиентов. Для этого придумали как раз **контекст**. Нужно теперь все брать из контекста внутри компонентов. И теперь нужно удалить **типизацию** массива ингредиентов в `BurgerIngredients` и `BurgerConstructor`

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -11,11 +11,41 @@ import {
+    (el) => el.type !== "bun"
+  );
+
+  const [orderState, orderDetailsDispatcher] = useReducer(
```
Хороший редьюсер сделали

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -58,10 +59,12 @@ function App() {
-          <BurgerIngredients ingredients={data} />
-          <BurgerConstructor ingredients={data} />
-        </main>
+        <IngredientsContext.Provider value={{ ingredients, setIngredients }}>
```
Хороший общий контекст сделали

gennady-bars commented on src/components/burger-constructor/burger-constructor-confirm.jsx:
```diff
@@ -24,11 +28,25 @@ function BurgerConstructorConfirm({ orderId }) {
+      setOrderId((currentState) => {
+        const newState = { ...currentState, id: res.order.number };
+        return newState;
+      })
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

**Нужно исправить это везде по коду в конце запросов

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -7,12 +7,13 @@ import IngredientsTab from "./ingredients-tab";
-  const mains = props.ingredients.filter((el) => el.type === "main");
+function BurgerIngredients() {
+  const { ingredients } = useContext(IngredientsContext);
+  const buns = ingredients.data.filter((el) => el.type === "bun");
```
**Можно лучше**


* Лучше ограничивать кол-во срабатываний фильтраций массивов при помощи хука `useMemo`.  Это касается таких операций, как `filter`, `map`, `forEach`, `reduce` и тд . Подробнее https://ru.reactjs.org/docs/hooks-reference.html#usememo

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Хороший редьюсер сделали
* Хороший общий контекст сделали

но есть некоторые недочеты: 


* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* `app.css` нужно сделать модулем `app.module.css`, а в компоненте использовать модульные классы, а не строки 
* В ` BurgerIngredients` и `BurgerConstructor` больше не нужно передавать в `props` массив ингредиентов. Для этого придумали как раз **контекст**. Нужно теперь все брать из контекста внутри компонентов. И теперь нужно удалить **типизацию** массива ингредиентов в `BurgerIngredients` и `BurgerConstructor`
* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.






**Можно лучше**


* Лучше ограничивать кол-во срабатываний фильтраций массивов при помощи хука `useMemo`.  Это касается таких операций, как `filter`, `map`, `forEach`, `reduce` и тд . Подробнее https://ru.reactjs.org/docs/hooks-reference.html#usememo 
* При кликах на табы `Булки, Соусы, Начинки` можно сделать логику **`плавного`** скролла до секции. Это можно сделать с помощью хука `useRef` и методом ` scrollIntoView`. Нужно установить для 3х секций 3 `useRef` и следить за изменением табов при кликах на них через функцию клика. Как меняется таб, то запускается `scrollIntoView ` со значением `{behavior: "smooth"}`
Подробнее про `useRef` тут https://ru.reactjs.org/docs/hooks-reference.html#useref
Подробнее про ` scrollIntoView` тут https://developer.mozilla.org/ru/docs/Web/API/Element/scrollIntoView
*  лучше не делать в **каждом** ингредиенте по модальному окну. Нужно было сделать одно окно для деталей в ` BurgerIngredients` (или `App`) и передавать в каждый ингредиент функцию **клика**, в которой бы назначались данные для показа в окне. Это поможет в спринте, где **роутинг** будете изучать и реализовывать



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 


Удачного рефакторинга кода.

piladik committed:
Поправил код в соответствии с замечаниями
1. Добавил базовый api
2. Добавил функцию request и заменил ею fetch
3. Добавил app.module.css
4. Удалил пропсы с BurgerIngredients и BurgerConstructor
5. Добавил catch блок в BurgerConstructorConfirm
6. Обернул фильтрации массивов в useMemo

**Клики на табы постараюсь сделать в конце спринта, если хватит времени
***С модальными окнами тоже постараюсь в конце спринта

gennady-bars commented on src/utils/burger-api.js:
```diff
@@ -1,33 +1,32 @@
   }
 
-  return 0;
+  return res;
```
нужно было сразу ответ распарсивать, чтобы не делать этого в каждом отдельном запросе

```js

  return res.json();

```

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
